File: backend/api/workflows.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.ext.asyncio import AsyncSession

from api.deps import get_current_user
from db.base import get_db
from schemas.workflow import (
    DeployResponse,
    StatusResponse,
    ValidationResult,
    WorkflowCreate,
    WorkflowResponse,
    WorkflowSummary,
    WorkflowTemplateResponse,
    WorkflowUpdate,
)
from services import workflow_service

router = APIRouter(prefix="/api/workflows", tags=["workflows"])


@router.get("/templates", response_model=list[WorkflowTemplateResponse])
async def list_workflow_templates():
    return workflow_service.get_workflow_templates()


@router.get("/templates/{slug}", response_model=WorkflowTemplateResponse)
async def get_workflow_template(slug: str):
    tmpl = workflow_service.get_workflow_template(slug)
    if not tmpl:
        raise HTTPException(status_code=404, detail="Workflow template not found")
    return tmpl


@router.get("", response_model=list[WorkflowSummary])
async def list_workflows(
    db: AsyncSession = Depends(get_db),
    current_user: dict = Depends(get_current_user),
):
    return await workflow_service.list_workflows(db, current_user["org_id"])


@router.post("", response_model=WorkflowResponse, status_code=201)
async def create_workflow(
    data: WorkflowCreate,
    db: AsyncSession = Depends(get_db),
    current_user: dict = Depends(get_current_user),
):
    return await workflow_service.create_workflow(db, data, current_user["org_id"], current_user["id"])


@router.get("/{workflow_id}", response_model=WorkflowResponse)
async def get_workflow(
    workflow_id: str,
    db: AsyncSession = Depends(get_db),
    current_user: dict = Depends(get_current_user),
):
    wf = await workflow_service.get_workflow(db, workflow_id)
    if not wf or wf.org_id != current_user["org_id"]:
        raise HTTPException(status_code=404, detail="Workflow not found")
    return wf


@router.patch("/{workflow_id}", response_model=WorkflowResponse)
async def update_workflow(
    workflow_id: str,
    data: WorkflowUpdate,
    db: AsyncSession = Depends(get_db),
    current_user: dict = Depends(get_current_user),
):
    wf = await workflow_service.get_workflow(db, workflow_id)
    if not wf or wf.org_id != current_user["org_id"]:
        raise HTTPException(status_code=404, detail="Workflow not found")
    try:
        return await workflow_service.update_workflow(db, wf, data)
    except ValueError as e:
        raise HTTPException(status_code=409, detail=str(e))


@router.delete("/{workflow_id}", status_code=204)
async def delete_workflow(
    workflow_id: str,
    db: AsyncSession = Depends(get_db),
    current_user: dict = Depends(get_current_user),
):
    wf = await workflow_service.get_workflow(db, workflow_id)
    if not wf or wf.org_id != current_user["org_id"]:
        raise HTTPException(status_code=404, detail="Workflow not found")
    try:
        await workflow_service.delete_workflow(db, wf)
    except ValueError as e:
        raise HTTPException(status_code=409, detail=str(e))


@router.get("/{workflow_id}/sessions")
async def list_sessions(
    workflow_id: str,
    db: AsyncSession = Depends(get_db),
    current_user: dict = Depends(get_current_user),
):
    from sqlalchemy import select
    from db.models import WorkflowSession
    wf = await workflow_service.get_workflow(db, workflow_id)
    if not wf or wf.org_id != current_user["org_id"]:
        raise HTTPException(status_code=404, detail="Workflow not found")
    result = await db.execute(
        select(WorkflowSession).where(WorkflowSession.workflow_id == workflow_id)
    )
    return result.scalars().all()


@router.post("/{workflow_id}/validate", response_model=ValidationResult)
async def validate_workflow(
    workflow_id: str,
    db: AsyncSession = Depends(get_db),
    current_user: dict = Depends(get_current_user),
):
    wf = await workflow_service.get_workflow(db, workflow_id)
    if not wf or wf.org_id != current_user["org_id"]:
        raise HTTPException(status_code=404, detail="Workflow not found")
    errors = workflow_service.validate_workflow_graph(wf.graph_definition)
    return ValidationResult(valid=len(errors) == 0, errors=errors)


@router.post("/{workflow_id}/deploy", response_model=DeployResponse)
async def deploy_workflow(
    workflow_id: str,
    db: AsyncSession = Depends(get_db),
    current_user: dict = Depends(get_current_user),
):
    wf = await workflow_service.get_workflow(db, workflow_id)
    if not wf or wf.org_id != current_user["org_id"]:
        raise HTTPException(status_code=404, detail="Workflow not found")
    try:
        return await workflow_service.deploy_workflow(db, wf)
    except ValueError as e:
        raise HTTPException(status_code=422, detail=str(e))


@router.post("/{workflow_id}/pause", response_model=StatusResponse)
async def pause_workflow(
    workflow_id: str,
    db: AsyncSession = Depends(get_db),
    current_user: dict = Depends(get_current_user),
):
    wf = await workflow_service.get_workflow(db, workflow_id)
    if not wf or wf.org_id != current_user["org_id"]:
        raise HTTPException(status_code=404, detail="Workflow not found")
    return await workflow_service.pause_workflow_action(db, wf)


@router.post("/{workflow_id}/resume", response_model=StatusResponse)
async def resume_workflow(
    workflow_id: str,
    db: AsyncSession = Depends(get_db),
    current_user: dict = Depends(get_current_user),
):
    wf = await workflow_service.get_workflow(db, workflow_id)
    if not wf or wf.org_id != current_user["org_id"]:
        raise HTTPException(status_code=404, detail="Workflow not found")
    return await workflow_service.resume_workflow_action(db, wf)

File: frontend/src/api/workflows.ts
import { apiClient } from './client'

export interface WorkflowNode {
  id: string
  type: string
  position: { x: number; y: number }
  data: Record<string, unknown>
}

export interface WorkflowEdge {
  id: string
  source: string
  target: string
  data?: Record<string, unknown>
}

export interface GraphDefinition {
  nodes: WorkflowNode[]
  edges: WorkflowEdge[]
}

export interface WorkflowSummary {
  id: string
  name: string
  template_slug: string | null
  status: string
  created_at: string
}

export interface Workflow extends WorkflowSummary {
  org_id: string
  graph_definition: GraphDefinition
  compiled_graph: Record<string, unknown> | null
  temporal_workflow_id: string | null
  created_by: string
  updated_at: string | null
}

export interface WorkflowTemplate {
  slug: string
  name: string
  description: string
  graph_definition: GraphDefinition
}

export interface ValidationResult {
  valid: boolean
  errors: string[]
}

export const workflowsApi = {
  listTemplates: () => apiClient.get<WorkflowTemplate[]>('/api/workflows/templates').then(r => r.data),
  getTemplate: (slug: string) => apiClient.get<WorkflowTemplate>(`/api/workflows/templates/${slug}`).then(r => r.data),
  list: () => apiClient.get<WorkflowSummary[]>('/api/workflows').then(r => r.data),
  get: (id: string) => apiClient.get<Workflow>(`/api/workflows/${id}`).then(r => r.data),
  create: (data: { name: string; template_slug?: string; graph_definition: GraphDefinition }) =>
    apiClient.post<Workflow>('/api/workflows', data).then(r => r.data),
  update: (id: string, data: { name?: string; graph_definition?: GraphDefinition }) =>
    apiClient.patch<Workflow>(`/api/workflows/${id}`, data).then(r => r.data),
  delete: (id: string) => apiClient.delete(`/api/workflows/${id}`),
  validate: (id: string) => apiClient.post<ValidationResult>(`/api/workflows/${id}/validate`).then(r => r.data),
  deploy: (id: string) => apiClient.post<{ status: string; temporal_workflow_id: string }>(`/api/workflows/${id}/deploy`).then(r => r.data),
  pause: (id: string) => apiClient.post<{ status: string }>(`/api/workflows/${id}/pause`).then(r => r.data),
  resume: (id: string) => apiClient.post<{ status: string }>(`/api/workflows/${id}/resume`).then(r => r.data),
  listSessions: (id: string) => apiClient.get(`/api/workflows/${id}/sessions`).then(r => r.data),
}
